workflows/check-cherry-picks: post review comments

wolfgangwalther opened the pull request:
Instead of failing the job, the workflow will now post review comments as "Request Changes". This makes the feedback more readily visible and avoids having to merge despite a failing CI job. It is also a pre-requisite to enable required status checks / required workflows in the future.

Committers are asked to confirm the differences by explicitly dismissing the generated review. After dismissal, the related review comment will automatically be marked as "resolved".

The comments only report warnings and errors. Reviews are automatically dismissed when they have been addressed by the author and no problems remain. If problems remain, existing, still pending, review comments will be updated. If the same problems had already been dismissed earlier, no new review comment will be created either.

Discussed in https://github.com/NixOS/nixpkgs/pull/411709#issuecomment-2916804988

Note: The full content of an example comment for this PR can be found in the summary page for the check-cherry-picks workflow. This is an example based on this PR, which is not a backport, so will only raise warnings for each commit, because there is no commit hash to pick. The workflow can't post the comment in this PR, because it runs in the `pull_request` context right now, which doesn't have the permissions to do so. Posting review comments will only work once the workflow runs in the `pull_request_target` context after merge.

The example doesn't contain a sample of the "please review the diff" case. In this case there will be a collapsable diff displayed in the warning, which is created as a github codeblock of type `diff`. This gives some basic colors for added/removed lines. In the summary page, the commit hashes are display as raw values, but in the comment GitHub will automatically turn them into links to the actual commits, which also allow hovering over them to see the commit message.

## Things done

- [x] Tested the script locally.
- [x] Tested the actions in my fork. 
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
ci/check-cherry-picks: improve formatting of failure reports

wolfgangwalther committed:
ci/check-cherry-picks: remove unused condition

The command substitution style we recently switched to strips trailing
newlines, so we don't need to check for empty lines anymore.

wolfgangwalther committed:
ci/check-cherry-picks: set -u

wolfgangwalther committed:
ci/check-cherry-picks: refactor output / logging

This way it's a bit more centralized and easier to extend.

wolfgangwalther committed:
workflows/check-cherry-picks: post review comments

Instead of failing the job, the workflow will now post review comments
as "Request Changes". This makes the feedback more readily visible and
avoids having to merge despite a failing CI job. It is also a
pre-requisite to enable required status checks / required workflows in
the future.

Committers are asked to confirm the differences by explicitly dismissing
the generated review. After dismissal, the related review comment will
automatically be marked as "resolved".

The comments only report warnings and errors. Reviews are automatically
dismissed when they have been addressed by the author and no problems
remain. If problems remain, existing, still pending, review comments
will be updated. If the same problems had already been dismissed
earlier, no new review comment will be created either.

JohnRTitor commented:
> the workflow will now post review comments as "Request Changes".

Sometimes we have to modify a commit, for conflict resolution.  This is bad, it'll block the PR, as commiters have the permission to "block" the PR by "requesting changes". nixpkgs-ci in the sense is also a commiter.

MattSturgeon reviewed (commented):
I always struggle to review (or develop) github actions/workflows with any confidence. But I don't see any obvious issues on an initial read-through of the patches.

MattSturgeon commented:
> > the workflow will now post review comments as "Request Changes".
> 
> Sometimes we have to modify a commit, for conflict resolution. This is bad, it'll block the PR, as commiters have the permission to "block" the PR by "requesting changes". nixpkgs-ci in the sense is also a commiter.
> 
> IMO, review "Comment" should be used instead.

I disagree. Anyone who can merge a PR can also dismiss a "Request changes" review. This being a manual step encourages the reviewer to actually check whether the review _should_ be dismissed or not.

![Screenshot From 2025-05-29 14-08-38](https://github.com/user-attachments/assets/3874d763-a0d7-4910-bd45-a4a1e0ce1632)
![Screenshot From 2025-05-29 14-08-52](https://github.com/user-attachments/assets/756b2d55-1088-40ba-b304-e1d44733e6de)

The generated review text explicitly mentions that committers should dismiss the review if it is invalid.

risicle commented:
So glad this is getting some more love

MattSturgeon commented on .github/workflows/reviews.yml:
```diff
@@ -0,0 +1,26 @@
+
+jobs:
+  cherry-pick:
+    name: Resolve dismissed CI reviews
```
```suggestion
    name: Minimize "dismissed" CI reviews as "resolved"
```

MattSturgeon commented on .github/workflows/reviews.yml:
Maybe this file needs some additional context, such as a comment explaining why it exists?

> The `check-cherry-picks` workflow creates review comments that should sometimes be manually dismissed.
> When a Ci-generated review is dismissed, this workflow automatically minimizes it, to prevent it from cluttering the PR.

MattSturgeon commented on .github/workflows/reviews.yml:
```diff
@@ -0,0 +1,26 @@
+name: "Reviews"
```
```suggestion
name: "Minimize dismissed reviews"
```

MattSturgeon commented on .github/workflows/reviews.yml:
```diff
@@ -0,0 +1,26 @@
+    steps:
+      - uses: actions/github-script@60a0d83039c74a4aee543508d2ffcb1c3799cdea # v7.0.1
+        with:
+          script: |
+            await github.graphql(`mutation($node_id:ID!) {
+              minimizeComment(input: {
+                classifier: RESOLVED,
+                subjectId: $node_id
+              })
+              { clientMutationId }
+            }`, { node_id: context.payload.review.node_id })
```
Does this actually need JS? Would the workflow be any more efficient by using the `gh` CLI tool to send the request instead?

```yml
- env:
    node_id: ${{ github.event.review.node_id }}
  run: |
    gh api graphql -F node_id="$node_id" -f query='
      mutation($node_id:ID!) {
        minimizeComment(input: {
          classifier: RESOLVED,
          subjectId: $node_id
        })
        { clientMutationId }
      }
    '
```

_(I don't think I converted the context path correctly, but as a rough example...)_

If it's just a style choice and there's no downside to either approach, feel free to ignore this comment!

Or maybe you prefer this style to accommodate future expansion of the workflow?

MattSturgeon replied:
Thanks for explaining!

philiptaron commented:
The attention to detail here is _astounding_, @wolfgangwalther. Very nice work.

Mic92 commented on ci/check-cherry-picks.sh:
```diff
@@ -75,23 +101,27 @@ while read -r new_commit_sha ; do
 
-          $range_diff_common --color
+          # First line contains commit SHAs, which we already printed.
+          $range_diff_common --color | tail -n +2
```
There is a limit how much text is allowed in comments.
You can look it up here: https://github.com/Mic92/nixpkgs-review/pull/491/files

MattSturgeon replied:
> You can look it up here: [Mic92/nixpkgs-review#491 (files)](https://github.com/Mic92/nixpkgs-review/pull/491/files)

> ```py
> # https://github.com/orgs/community/discussions/27190
> MAX_GITHUB_COMMENT_LENGTH = 65536
> ```

> Should we handle this here or let it fail instead?

Can we use `wc` to test whether to include the diff in the comment or not, but still print it to the CI logs regardless?

It's probably not the end of the world if it fails, since that'd make the CI job fail and also block the PR. At least until requiring status checks befor merge.

But it's not the ideal DX.

> So I'd say, we let it fail for now and see where that goes?

I think it's reasonable to be left for followup work, maybe highlight it as a TODO comment and/or a note on the PR description?

Mic92 replied:
For normal code changes I don't expect that we would hit the limit. However we also have sometimes large lock file changes that may fall into this category.

wolfgangwalther committed:
workflows/check-cherry-picks: truncate long diffs after 10k characters

GitHub comments have a length limit, so we can't just dump everything.
The 10k limit is arbitrary, but the assumption is that reviewing the
range-diff is not the sensible thing to do once it becomes a certain
size - reviewing the regular diff and treating the commit as "new" is
easier to do in that case. Thus, truncating should work out fine,
especially when the full range-diff is still available in the runner
log.

This could still end up in with an error, if a PR has multiple commits,
which all hit the limit. Let's get there first, before we try to fix
that hypothetical case, too.

philiptaron commented on .github/workflows/check-cherry-picks.yml:
```diff
@@ -24,8 +25,98 @@ jobs:
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              pull_number: context.payload.pull_request.number,
+              per_page: 100
```
Ugh, that's really quite irritating. It looks based on reading https://docs.github.com/en/rest/using-the-rest-api/using-pagination-in-the-rest-api?apiVersion=2022-11-28 that the loop is seeing if there's a link, then making another request with the current request's linked rel="next". 

No idea how that shows up in octokit.js. There's also the page parameter, which seems like it might be relevant.

philiptaron commented on .github/workflows/check-cherry-picks.yml:
```diff
@@ -24,8 +25,105 @@ jobs:
+          script: |
+            const { readFile, writeFile } = require('node:fs/promises')
+
+            const job_url = (await github.paginate(github.rest.actions.listJobsForWorkflowRun, {
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              run_id: context.runId
+            }))[0].html_url + '?pr=' + context.payload.pull_request.number
```
Because this might return multiple things, is mapping over it to make an array the better choice instead of `[0]`?

It _should_ return one thing, I think. Or if it returns multiple things, they've all got the same `html_url`.

philiptaron reviewed (approved):
Love the pagination!

MattSturgeon commented on ci/check-cherry-picks.sh:
```diff
@@ -75,23 +101,37 @@ while read -r new_commit_sha ; do
+          max_length=10000
+          truncated="$($range_diff_common | tail -n +2 | sed -Ee 's/^ {4}/> /g' \
+            | head -c $max_length)"
+          if [ "${#truncated}" -ge $max_length ]; then
```
Should this check the actual diff size instead of the truncated size?

1 in a million edge case where the two sizes happen to be the same, so we erroneously append the "...truncated..." suffix?

I guess this doesn't really matter.

MattSturgeon commented on ci/check-cherry-picks.sh:
```diff
@@ -75,23 +101,37 @@ while read -r new_commit_sha ; do
+          # The output of `git range-diff` is indented with 4 spaces, which we need to match with the
+          # code blocks indent to get proper syntax highlighting on GitHub.
+          # Also limit the output to 10k bytes (and remove the last, potentially incomplete line), because
+          # GitHub comments are limited in length. The value of 10k is arbitrary with the assumption, that
+          # after the range-diff becomes a certain size, a reviewer is better off reviewing the regular diff
+          # in GitHub's UI anyway, thus treating the commit as "new" and not cherry-picked.
```
+1 for highlighting that the limit is arbitrary. I guess in theory, a PR with a number of large individual commits could still hit GitHub's limit, despite _each_ diff being limited to 10k. Should this also get a brief mention?

MattSturgeon commented on .github/workflows/check-cherry-picks.yml:
```diff
@@ -24,8 +25,105 @@ jobs:
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              pull_number: context.payload.pull_request.number
+            })).find(review =>
+              review.user.login == 'github-actions[bot]' && (
+                // If a review is still pending, we can just update this instead
+                // of posting a new one.
+                review.state == 'CHANGES_REQUESTED' ||
+                // No need to post a new review, if an older one with the exact
+                // same content had already been dismissed.
+                review.body == body
+              )
+            )
```
As `find` returns the first match, could it happen that an old review is dismissed, and new changes come along that update the review, but it incorrectly stays dismissed?

Also, is it possible to get into a state where there are multiple review comments (e.g. one is dismissed, then another different one is posted)? In that scenario I think this `find` might find the wrong one.

Just trying to wrap my head around whether there are any weird states we could get into, where our assumptions are broken regarding which review comments should be updated.

Maybe I'm just looking at it too hard; it looked fine the last three times I read it 😅

MattSturgeon replied:
Dismissed reviews no longer count as "requesting changes", so I think that rules out some of the weird states I was imagining.

MattSturgeon reviewed (approved):
Nice job! 👏
